### PR TITLE
[iOS] Fix App crash when use WebView with iOS VoiceOver

### PR DESCRIPTION
**Pull request by tacck**
This PR fix the issue [flutter/flutter #140528](https://github.com/flutter/flutter/issues/140528).

This fix cleans up declarations regarding `retain` and `release`.
`flutterAccessibilityContainer` needs to be managed in `FlutterPlatformViews.mm`, and problems will occur if `retain` or `release` is performed in `SemanticsObject.mm`.

This fix is checked by the `testFlutterPlatformViewSemanticsContainer` function in the `SemanticsObjectTest.mm`.
The result is not changed.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chinmaygarde** [comment]
cc @jmagman

**Review by jmagman — COMMENTED** [review]
> This fix is checked by the `testFlutterPlatformViewSemanticsContainer` function in the `SemanticsObjectTest.mm`.
> The result is not changed.

Are you able to get a test to fail/crash on main but succeed on this PR, that actually exercises the bad behavior?  Your fix looks more correct, but in practice I can't see how it's actually changing the retains in a way that would prevent a crash.

This test, for example, passed on both main and this PR:
```objc
- (void)testFlutterPlatformViewSemanticsContainer {
  fml::WeakPtrFactory<flutter::testing::MockAccessibilityBridge> factory(
      new flutter::testing::MockAccessibilityBridge());
  fml::WeakPtr<flutter::testing::MockAccessibilityBridge> bridge = factory.GetWeakPtr();
  __weak FlutterTouchInterceptingView* weakPlatformView;
  __weak FlutterPlatformViewSemanticsContainer* weakContainer;
  @autoreleasepool {
    FlutterTouchInterceptingView* platformView = [[FlutterTouchInterceptingView alloc] init];
    weakPlatformView = platformView;

    @autoreleasepool {
      FlutterPlatformViewSemanticsContainer* container =
      [[FlutterPlatformViewSemanticsContainer alloc] initWithBridge:bridge
                                                                uid:1
                                                       platformView:platformView];
      weakContainer = container;
      XCTAssertEqualObjects(platformView.accessibilityContainer, container);
      XCTAssertNotNil(weakPlatformView);
      XCTAssertNotNil(weakContainer);
    }
    XCTAssertNotNil(weakPlatformView);
    XCTAssertNil(weakContainer);
  }
  // Check if there's no more strong references to `platformView` after container and platformView
  // are released.
  XCTAssertNil(weakPlatformView);
  XCTAssertNil(weakContainer);
}
```

By the way, I'm hoping to migrate `FlutterPlatformViews` to ARC soon to prevent these kinds of issues 🙂

**Comment by tacck** [comment]
@jmagman 
Thank you for your review.
I modified the test code based on the example you wrote.
https://github.com/flutter/engine/pull/52484/files#diff-31e401d2bb8996a130877ac6c34512d746f122368d40946914a592c1648a7f07

> Are you able to get a test to fail/crash on main but succeed on this PR, that actually exercises the bad behavior? Your fix looks more correct, but in practice I can't see how it's actually changing the retains in a way that would prevent a crash.

This test verifies that the `container` object continues to exist even if it moves to a different scope.
This is because the `platformView` retains the `container`.
When `platformView` is destroyed, `container` will also be destroyed simultaneously.
It is the difference between main and PR.

> By the way, I'm hoping to migrate `FlutterPlatformViews` to ARC soon to prevent these kinds of issues 🙂

That's great.
I hope it goes well 😊

**Comment by jmagman** [comment]
> This test verifies that the `container` object continues to exist even if it moves to a different scope. This is because the `platformView` retains the `container`. When `platformView` is destroyed, `container` will also be destroyed simultaneously. It is the difference between main and PR.

Hm, this test also passes on main though.

**Comment by tacck** [comment]
I validated my fixes and tests using the following process.
Please let me know if there are any issues.

---

**Download tools and add PATH.**
[SETTING UP](https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/depot_tools_tutorial.html#_setting_up)
```
git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
export PATH=${PATH}:$(pwd)/depot_tools
```

**Download the code and prepare.**
[Setting up the Engine development environment](https://github.com/flutter/flutter/wiki/Setting-up-the-Engine-development-environment)
```
mkdir engine; cd engine;
fetch flutter
git -C src/flutter remote rename origin upstream
gclient sync
```

git pull again at 2024/05/09 11:00+09:00 (2024/05/09 02:00+00:00 UTC)
```
git -C src/flutter pull upstream main
gclient sync
```

**Build engine.**
[Compiling the engine](https://github.com/flutter/flutter/wiki/Compiling-the-engine)
```
cd src
./flutter/tools/gn --ios --unoptimized --simulator --simulator-cpu=arm64
ninja -C out/ios_debug_sim_unopt_arm64
./flutter/tools/gn --unoptimized --mac-cpu arm64
ninja -C out/host_debug_unopt_arm64
```

**Do test.**
[Testing the engine](https://github.com/flutter/flutter/wiki/Testing-the-engine)
```
./flutter/testing/run_tests.py --type=objc --ios-variant=ios_debug_sim_unopt_arm64 --variant=host_debug_unopt_arm64
```

Result:
```
(snip)
** TEST SUCCEEDED **
(snip)
```

Now, I think the environment is set up.


**Patch the test code and do the below again.**
```
./flutter/tools/gn --ios --unoptimized --simulator --simulator-cpu=arm64
ninja -C out/ios_debug_sim_unopt_arm64
./flutter/tools/gn --unoptimized --mac-cpu arm64
ninja -C out/host_debug_unopt_arm64
./flutter/testing/run_tests.py --type=objc --ios-variant=ios_debug_sim_unopt_arm64 --variant=host_debug_unopt_arm64
```

Result:
```
(snip)
Test Case '-[SemanticsObjectTest testFlutterPlatformViewSemanticsContainer]' started.
/../../flutter/shell/platform/darwin/ios/framework/Source/SemanticsObjectTest.mm:1015: error: -[SemanticsObjectTest testFlutterPlatformViewSemanticsContainer] : ((weakContainer) != nil) failed
Test Case '-[SemanticsObjectTest testFlutterPlatformViewSemanticsContainer]' failed (0.058 seconds).
(snip)
** TEST FAILED **
(snip)
```

This means that the test code catches the checkpoint in the main branch, I think.


**Patch all code and do above again.**

Result:
```
** TEST SUCCEEDED **
```

This means that the product code fix the probrem, I think.

**Review by jmagman — APPROVED** [review]
Apologies for wasting your time, you're totally right, I was running the test incorrectly (I was playing around with `bin/et` and it output to `out/ci/ios_debug_unopt_sim_arm64` instead of `out/ios_debug_sim_unopt` 😓 

https://github.com/flutter/engine/blob/5eaa296fe5eef467bc694f1ee990eb9326540a35/testing/ios/IosUnitTests/Tests/FlutterEngineConfig.xcconfig#L1-L3
)

LGTM, thanks for the fix!

**Comment by jmagman** [comment]
+ @hellohuanlin for second review.

File: shell/platform/darwin/ios/framework/Source/SemanticsObject.mm [hunk 1]
```diff
@@ -876,14 +876,13 @@ - (instancetype)initWithBridge:(fml::WeakPtr<flutter::AccessibilityBridgeIos>)br
...
-    _platformView = [platformView retain];
+    _platformView = platformView;
     [platformView setFlutterAccessibilityContainer:self];
   }
   return self;
 }
 
 - (void)dealloc {
-  [_platformView release];
   _platformView = nil;
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:886]: nit: it's not necessary to nil out an ivar here (can remove the whole dealloc).

**Review by hellohuanlin — APPROVED** [review]
Good catch! Thanks for digging into this.

File: shell/platform/darwin/ios/framework/Source/SemanticsObject.mm (hunk 1 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:886]: I'm about to remove this dealloc in https://github.com/flutter/engine/pull/52535, I wouldn't worry about it in this PR.

Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:886]: Oh actually that's not true, the PR doesn't include `SemanticsObject`.  But I'll get to that one hopefully soon 🙂

**Comment by agacemi** [comment]
@zanderso any idea if this fix will be released as a patch of the current release 3.22 !

**Comment by jmagman** [comment]
This fix is in 3.23.0-10.0.pre which hasn't reached the beta channel yet.  See https://github.com/flutter/flutter/blob/master/docs/releases/Hotfixes-to-the-Stable-Channel.md

**Comment by tacck** [comment]
@JDongKhan 
This PR was incorporated into 3.24.0.
https://docs.flutter.dev/release/release-notes/release-notes-3.24.0

**Pull request merged**